[clang] Use separator for large numeric values in overflow diagnostic

Atousa opened the pull request:
Add functionality to APInt::toString() that allows it to insert
separators between groups of digits, using the C++ litteral
separator ' between groups.
    
Fixes issue #58228

Reviewers :  @AaronBallman, @cjdb, @tbaederr

Atousa committed:
[clang] Use separator for large numeric values in overflow diagnostic

Add functionality to APInt::toString() that allows it to insert
separators between groups of digits, using the C++ litteral
separator ' between groups.

Fixes issue #58228

AaronBallman commented:
Overall, I think this is a *fantastic* idea for diagnostic readability, thank you for the patch! Please be sure to add a release note so users known to expect the changes. The Windows precommit CI failure is unrelated and can be ignored.

Once you think you're ready, be sure to move the PR out of draft status so reviewers remember to give it a proper review. My initial glance didn't spot any issues though.

Atousa committed:
Merge branch 'main' into main

Atousa committed:
Update clang/lib/AST/ExprConstant.cpp

Co-authored-by: Timm Baeder <tbaeder@redhat.com>

Atousa committed:
Update llvm/lib/Support/APInt.cpp

Co-authored-by: Timm Baeder <tbaeder@redhat.com>

Atousa committed:
Update llvm/lib/Support/APInt.cpp

Co-authored-by: Timm Baeder <tbaeder@redhat.com>

Atousa committed:
Merge branch 'main' into main

Atousa committed:
Update printing APInt numbers with separators

The new constant interpreter can also emit integer constant overflow
warnings.  The code was updated to also print these with separators.

tbaederr commented on llvm/lib/Support/APInt.cpp:
```diff
@@ -2187,6 +2188,12 @@ void APInt::toString(SmallVectorImpl<char> &Str, unsigned Radix, bool Signed,
 
+  // Number of digits in a group between separators.
+  unsigned Grouping = (Radix == 8 || Radix == 10) ? 3 : 4;
+  if (Radix == 8 || Radix == 10) {
+    Grouping = 3;
+  }
+
```
This code block is now unnecessary.

Atousa commented:
Humble ping @cjdb 
Do you have any further comments? @AaronBallman @tbaederr

Atousa commented:
@AaronBallman @tbaederr, Do you have any further comments?

AaronBallman commented on llvm/include/llvm/ADT/APInt.h:
```diff
@@ -1626,7 +1626,8 @@ class [[nodiscard]] APInt {
   void toString(SmallVectorImpl<char> &Str, unsigned Radix, bool Signed,
-                bool formatAsCLiteral = false, bool UpperCase = true) const;
+                bool formatAsCLiteral = false, bool UpperCase = true,
+                bool insertSeparators = false) const;
```
```suggestion
                bool InsertSeparators = false) const;
```
This matches our usual coding style and is the prevailing style in the function signature (a follow-up NFC change could rename `formatAsCLiteral`).

AaronBallman commented on llvm/include/llvm/ADT/StringExtras.h:
```diff
@@ -329,9 +329,10 @@ inline std::string itostr(int64_t X) {
 inline std::string toString(const APInt &I, unsigned Radix, bool Signed,
-                            bool formatAsCLiteral = false) {
+                            bool formatAsCLiteral = false,
+                            bool upperCase = true, bool addSeparators = false) {
```
Same suggestion here as above (amusing that `upperCase` changed styles...).

AaronBallman commented on llvm/lib/Support/APInt.cpp:
```diff
@@ -2161,7 +2161,8 @@ void APInt::fromString(unsigned numbits, StringRef str, uint8_t radix) {
 void APInt::toString(SmallVectorImpl<char> &Str, unsigned Radix, bool Signed,
-                     bool formatAsCLiteral, bool UpperCase) const {
+                     bool formatAsCLiteral, bool UpperCase,
+                     bool insertSeparators) const {
```
And here too!

AaronBallman commented on llvm/lib/Support/APInt.cpp:
```diff
@@ -2257,17 +2266,27 @@ void APInt::toString(SmallVectorImpl<char> &Str, unsigned Radix, bool Signed,
+    int Pos = 0;
     while (Tmp.getBoolValue()) {
       unsigned Digit = unsigned(Tmp.getRawData()[0]) & MaskAmt;
+      if (insertSeparators && Pos % Grouping == 0 && Pos > 0) {
+        Str.push_back('\'');
+      }
```
Same here as above

AaronBallman commented on llvm/lib/Support/APInt.cpp:
```diff
@@ -2257,17 +2266,27 @@ void APInt::toString(SmallVectorImpl<char> &Str, unsigned Radix, bool Signed,
       uint64_t Digit;
       udivrem(Tmp, Radix, Tmp, Digit);
       assert(Digit < Radix && "divide failed");
+      if (insertSeparators && Pos % Grouping == 0 && Pos > 0) {
+        Str.push_back('\'');
+      }
```
Here too

AaronBallman commented on llvm/unittests/ADT/APIntTest.cpp:
```diff
@@ -1379,6 +1379,23 @@ TEST(APIntTest, toString) {
+  APInt(64, 1000000).toString(S, 16, false, true, true, true);
+  EXPECT_EQ(std::string(S), "0xF'4240");
+  S.clear();
+  APInt(64, 1'000'000'000).toString(S, 36, false, false, false, true);
+  EXPECT_EQ(std::string(S), "gj'dgxs");
```
lol, I love it (it took me a minute when I saw the expected string).

AaronBallman reviewed (commented):
You should also add a release note to `clang/docs/ReleaseNotes.rst` to tell users about the improved diagnostic behavior.

Aside from some small nits, this looks really close to ready!

Atousa committed:
[clang][NFC] Update code style  conventions

Atousa committed:
[clang][NFC] Update code style conventions

Atousa committed:
Merge branch 'main' into main

Atousa commented:
please @AaronBallman

AaronBallman commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -175,6 +175,9 @@ Improvements to Clang's diagnostics
    of ``-Wconversion``. Fixes `#69444 <https://github.com/llvm/llvm-project/issues/69444>`_.
 
+- Clang now uses thousand separators when printing large numbers in integer overflow diagnostics.
+  Fixes `#80939 <https://github.com/llvm/llvm-project/issues/80939>`_.
```
```suggestion
  Fixes #GH80939.
```
We just landed a change this morning to automagically expand GitHub issue numbers, so this isn't a functional change.

AaronBallman commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -2778,7 +2778,9 @@ static bool CheckedIntArithmetic(EvalInfo &Info, const Expr *E,
                                        diag::warn_integer_constant_overflow)
-          << toString(Result, 10) << E->getType() << E->getSourceRange();
+          << toString(Result, 10, Result.isSigned(), /*formatAsCLiteral=*/false,
+                      /*upperCase=*/true, /*insertSeparators=*/true)
```
```suggestion
                      /*UpperCase=*/true, /*InsertSeparators=*/true)
```
Same changes needed elsewhere, but this should keep clang-tidy happy for folks using it.

Atousa committed:
[clang][NFC] Update code style conventions

AaronBallman reviewed (approved):
LGTM, thank you for the improvement!
